instrumentor/controllers/instrumentationdevice/manager.go

```diff
@@ -3,10 +3,62 @@ package instrumentationdevice
 import (
 	odigosv1 "github.com/odigos-io/odigos/api/odigos/v1alpha1"
 	k8sutils "github.com/odigos-io/odigos/k8sutils/pkg/client"
+	appsv1 "k8s.io/api/apps/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 )
 
+type workloadEnvChangePredicate struct {
+	predicate.Funcs
+}
+
+func (w workloadEnvChangePredicate) Create(e event.CreateEvent) bool {
+	return false
+}
+
+func (w workloadEnvChangePredicate) Update(e event.UpdateEvent) bool {
+
+	if e.ObjectOld == nil || e.ObjectNew == nil {
+		return false
+	}
+
+	oldPodSpec, err := getPodSpecFromObject(e.ObjectOld)
+	if err != nil {
+		return false
+	}
+	newPodSpec, err := getPodSpecFromObject(e.ObjectNew)
+	if err != nil {
+		return false
+	}
+
+	// only handle workloads if any env changed
+	if len(oldPodSpec.Spec.Containers) != len(newPodSpec.Spec.Containers) {
+		return true
+	}
+	for i := range oldPodSpec.Spec.Containers {
+		if len(oldPodSpec.Spec.Containers[i].Env) != len(newPodSpec.Spec.Containers[i].Env) {
+			return true
+		}
+		for j := range oldPodSpec.Spec.Containers[i].Env {
+			if oldPodSpec.Spec.Containers[i].Env[j].Value != newPodSpec.Spec.Containers[i].Env[j].Value {
+				return true
```

Comment:
Do we need to check the env name before checking the value?

Reply:
good point, thanks! updated

Reply:
I still think this can be improve to only check if the env vars we care about have changed/added/deleted